worlds/pokemon_emerald/rules.py

```diff
@@ -0,0 +1,1366 @@
+"""
+Logic rule definitions for Pokemon Emerald
+"""
+from BaseClasses import CollectionState, MultiWorld
+from Options import Toggle
+
+from worlds.generic.Rules import add_rule, set_rule
+
+from .options import EliteFourRequirement, NormanRequirement, ExtraBoulders
+from .util import location_name_to_label
+
+
+def _can_cut(state: CollectionState, player: int) -> bool:
+    return state.has("HM01 Cut", player) and state.has("Stone Badge", player)
+
+
+def _can_surf(state: CollectionState, player: int) -> bool:
+    return state.has("HM03 Surf", player) and state.has("Balance Badge", player)
+
+
+def _can_strength(state: CollectionState, player: int) -> bool:
+    return state.has("HM04 Strength", player) and state.has("Heat Badge", player)
+
+
+def _can_flash(state: CollectionState, player: int) -> bool:
+    return state.has("HM05 Flash", player) and state.has("Knuckle Badge", player)
+
+
+def _can_rock_smash(state: CollectionState, player: int) -> bool:
+    return state.has("HM06 Rock Smash", player) and state.has("Dynamo Badge", player)
+
+
+def _can_waterfall(state: CollectionState, player: int) -> bool:
+    return state.has("HM07 Waterfall", player) and state.has("Rain Badge", player)
+
+
+def _can_dive(state: CollectionState, player: int) -> bool:
+    return state.has("HM08 Dive", player) and state.has("Mind Badge", player)
+
+
+def _can_use_mach_bike(state: CollectionState, player: int) -> bool:
+    return state.has("Mach Bike", player)
+
+
+def _can_use_acro_bike(state: CollectionState, player: int) -> bool:
+    return state.has("Acro Bike", player)
+
+
+def _defeated_n_gym_leaders(state: CollectionState, player: int, n: int) -> bool:
+    num_gym_leaders_defeated = 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_ROXANNE", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_BRAWLY", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_WATTSON", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_FLANNERY", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_NORMAN", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_WINONA", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_TATE_AND_LIZA", player) else 0
+    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_JUAN", player) else 0
+    return num_gym_leaders_defeated >= n
+
+
+# Rules are organized by town/route/dungeon and ordered approximately
+# by when you would first reach that place in a vanilla playthrough.
+def set_default_rules(multiworld: MultiWorld, player: int) -> None:
+    can_cut = lambda state: _can_cut(state, player)
```

Comment:
you don't need these stand-ins, you can just assign the original methods

Reply:
I need to pass `player` to `state.has`. I could do
```py
def set_default_rules(multiworld: MultiWorld, player: int) -> None:
    can_cut = lambda state: state.has("HM01 Cut", player) and state.has("Stone Badge", player)
```
and duplicate that line for relevant HMs in `set_overworld_item_rules` and others. The point of `_can_cut` is just to reduce duplicated code, but they're simple enough that it shouldn't matter that much if you think it looks cleaner. I think it creates the same number of access rule functions.

Reply:
Ah right, that was the main reason i used a custom class in order to do that. the way you're doing it is fine, it just looks weird 🤷 .